src/python/grpcio/grpc/experimental/aio/_interceptor.py

```diff
@@ -0,0 +1,345 @@
+    started and the execution is still held by the interceptors or when the
+    RPC has finished but again the execution is still held by the interceptors.
+
+    Once the RPC is finally executed, all methods are finally done against the
+    intercepted call, being at the same time the same call returned to the
+    interceptors.
+
+    For most of the methods, like `initial_metadata()` the caller does not need
+    to wait until the interceptors task is finished, once the RPC is done the
+    caller will have the freedom for accessing to the results.
+
+    For the `__await__` method is it is proxied to the intercepted call only when
+    the interceptor task is finished.
+    """
+
+    _loop: asyncio.AbstractEventLoop
+    _channel: cygrpc.AioChannel
+    _cancelled_before_rpc: bool
+    _intercepted_call: Optional[_base_call.UnaryUnaryCall]
+    _intercepted_call_created: asyncio.Event
+    _interceptors_task: asyncio.Task
+
+    def __init__( # pylint: disable=R0913
+            self,
+            interceptors: Sequence[UnaryUnaryClientInterceptor],
+            request: RequestType,
+            timeout: Optional[float],
+            channel: cygrpc.AioChannel,
+            method: bytes,
+            request_serializer: SerializingFunction,
+            response_deserializer: DeserializingFunction) -> None:
+        self._channel = channel
+        self._loop = asyncio.get_event_loop()
+        self._interceptors_task = asyncio.ensure_future(
+            self._invoke(interceptors, method, timeout, request,
+                         request_serializer, response_deserializer))
+
+    def __del__(self):
+        self.cancel()
+
+    async def _invoke(
+            self, interceptors: Sequence[UnaryUnaryClientInterceptor],
+            method: bytes, timeout: Optional[float], request: RequestType,
+            request_serializer: SerializingFunction,
+            response_deserializer: DeserializingFunction) -> UnaryUnaryCall:
+        """Run the RPC call wrapped in interceptors"""
+
+        async def _run_interceptor(
+                interceptors: Iterator[UnaryUnaryClientInterceptor],
+                client_call_details: ClientCallDetails,
+                request: RequestType) -> _base_call.UnaryUnaryCall:
+            try:
+                interceptor = next(interceptors)
+            except StopIteration:
+                interceptor = None
+
+            if interceptor:
+                continuation = functools.partial(_run_interceptor, interceptors)
+                try:
+                    call_or_response = await interceptor.intercept_unary_unary(
+                        continuation, client_call_details, request)
+                except grpc.RpcError as err:
+                    # gRPC error is masked inside an artificial call,
+                    # caller will see this error if and only
+                    # if runs an `await call` operation
+                    return UnaryUnaryCallRpcError(err)
```

Comment:
Why do we need to mask the `grpc.RpcError` as `UnaryUnaryCallRpcError`? Instead, if the motivation is to let other interceptors able to handle the `RpcError`, we could simply let the exception propagate. You are using recursion here, the stack should do the trick for us.

Reply:
I'm doing this for the sake of consistency, remember that the pattern that we are prescribing for the caller is the following one:

```python
call = mutlicallable(...)
try:
    await call
except grpc.RpcError as err:
    print(await call.code())
except asyncio.CancelledError:
    print(await call.code())
```

Basically we are asking for following the next rules:

1 - The way of raising any issue is by awaiting the call, this would let you know if there was an error by raising an exception or returning a response if everything went fine.
2 - Even raising an exception call will keep and could be used for gathering the RPC attributes
3- This is especially relevant when an `asyncio.CancelledErrorr` happens, where the only place where RPC attributes can be gathered is by using the call

By masking the the `grpc.RpcErrror` and using a `UnaryUnaryCallRpcError` instead, we are making the previous pattern explicit in the interceptors side.

Without doing this the rules that we already set up for the caller could be easily unfollowed by the interceptors, so ending up with something like this:

```python
try:
    call = await continuation(....)
except grpc.RpcError as err:
    print(await call.code())
except asyncio.CancelledError:
    print(await call.code())
```

This code won't be following the prescription that we made for the caller, since the call will never be awaited, and in case of an error, the `call` will never get any value since the exception happened during the continuation execution.

So if the interceptor needs to take care of the outcome of a call this would need to follow the following pattern:

```python
try:
    call = await continuation(....)
    await call
except grpc.RpcError as err:
    print(await call.code())
except asyncio.CancelledError:
    print(await call.code())
```

Remember that the first await when the continuation is called only relates to the coroutine call, so is also needed a second await on the call returned. Hence, is when the except statements make sense since the call would have all of the RPC context.

Reply:
> 1 - The way of raising any issue is by awaiting the call, this would let you know if there was an error by raising an exception or returning a response if everything went fine.
> 2 - Even raising an exception call will keep and could be used for gathering the RPC attributes
> 3- This is especially relevant when an asyncio.CancelledErrorr happens, where the only place where RPC attributes can be gathered is by using the call

Yes. These rules are ones we are following for normal calls.

> This code won't be following the prescription that we made for the caller, since the call will never be awaited, and in case of an error, the call will never get any value since the exception happened during the continuation execution.

Good argument! This is indeed a problem I neglected. I think current solution is self-contained enough. Before I LGTM, I still want to have a discussion about another alternative:

Can we advice users to only access RPC's error status through `RpcError`?

I felt like most of the pain comes from a design that is too idealistic, which we are trying to support both `asyncio.CancelledError` and `RpcError` with `CANCELLED` status code. Since the `cancel` method is implemented by ourselves, we can make sure the `asyncio.CancelledError` is handled in the Cython layer and then we can wrap it with `RpcError`. So, users can only observe `RpcError` being raised from `await call`.

In this way, the error-handling section of code is unified in all scenarios.

```Python
# Normal Unary Call
try:
    response = await call
except grpc.RpcError as err:
    print(err.code())

# Interceptor only await continuation
try:
    call = await continuation(....)
except grpc.RpcError as err:
    print(err.code())

# Interceptor awaits RPC
try:
    call = await continuation(....)
    await call
except grpc.RpcError as err:
    print(err.code())
```

Reply:
Finally, I will follow your advice. So not masking the `RpcError` and `CancelledError` exceptions as fake calls. Could be some down-side effects of doing so, especially for the use case of the `CancelledError` where implicit task cancellations could remain unnoticeable at interceptor, as the following example shows:

```python
call = continuation(...)
await a_long_internal_asynchronous_coroutine()
....
```
The `a_long_internal_asynchronous_coroutine` would never be affected by a 
`CancelledError` exception, so not giving up on this execution even thought a cancellation was propagated by the down layers.

IMO this could be a good reason for not masking the `CancelledError`, the same for the
`RpcError`.

So, in that case, as you already mentioned we would need to advise the users for expecting both exceptions at continuation call and when the _call_ is being awaited.

Reply:
AsyncIO users should be prepared to observe `CancelledError` from awaited functions. Looking forward to see your code changes.